wdl-analysis/src/scope/v1.rs

```diff
@@ -175,704 +201,1013 @@ fn call_conflict(name: &Ident, first: NameContext, suggest_fix: bool) -> Diagnos
+                        ));
+                    } else {
+                        graph.add_edge(to, from, ());
+                    }
+                }
+            }
         }
+    }
 
-        // Ensure the import has a matching WDL version
-        let import_document = import_node.document().expect("import should have parsed");
-        let import_scope = import_node
-            .analysis()
-            .map(|a| a.scope().clone())
-            .expect("import should have been analyzed");
-
-        // Check for compatible imports
-        match import_document.version_statement() {
-            Some(stmt) => {
-                let our_version = stmt.version();
-                if matches!((our_version.as_str().split('.').next(), importer_version.as_str().split('.').next()), (Some(our_major), Some(their_major)) if our_major != their_major)
-                {
-                    return Err(Some(incompatible_import(
-                        our_version.as_str(),
-                        span,
-                        importer_version,
-                    )));
+    // At this point the graph is guaranteed acyclic; now calculate the struct types
+    // in topological order
+    for index in toposort(&graph, Some(&mut space)).expect("graph should be acyclic") {
+        let definition =
+            StructDefinition::cast(SyntaxNode::new_root(document.structs[index].node.clone()))
+                .expect("node should cast");
+
+        let structs = &document.structs;
+        let mut converter = AstTypeConverter::new(&mut document.types, |name, span| {
+            if let Some(s) = structs.get(name) {
+                Ok(s.ty().unwrap_or(Type::Union))
+            } else {
+                diagnostics.push(unknown_type(
+                    name,
+                    Span::new(span.start() + structs[index].offset, span.len()),
+                ));
+                Ok(Type::Union)
+            }
+        });
+
+        let ty = converter
+            .convert_struct_type(&definition)
+            .expect("struct type conversion should not fail");
+
+        let s = &mut document.structs[index];
+        assert!(s.ty.is_none(), "type should not already be present");
+        s.ty = Some(document.types.add_struct(ty));
+    }
+}
+
+/// Performs type checking on a document.
+fn type_check(document: &mut DocumentScope, ast: &Ast, diagnostics: &mut Vec<Diagnostic>) {
+    for item in ast.items() {
+        match item {
+            DocumentItem::Import(_) | DocumentItem::Struct(_) => continue,
+            DocumentItem::Task(definition) => {
+                if let Some(task) = document.tasks.get(definition.name().as_str()) {
+                    type_check_task(document, &definition, task.scope, diagnostics);
                 }
             }
-            None => {
-                return Err(Some(import_missing_version(span)));
+            DocumentItem::Workflow(_) => {
+                // TODO: implement
```

Comment:
Just a note that this is still TODO as well.

Reply:
I'll file a tracking issue to complete static analysis by implementing workflow type checks.

Reply:
Opened https://github.com/stjude-rust-labs/wdl/issues/167.